Signature verification of models based on a security policy

stefanberger opened the pull request:
<!-- markdownlint-disable -->

## Purpose

This series implements support for signature verification of AI models and LoRA adapters based on a vLLM security policy. The security policy is a JSON document that describes the AI models by their local path (later on to be extended to Huggingface repo path) along with their expected signers. Basing signature verification on a policy seems more flexible than relying on command line parameters as previously attempted in PR #21957. This is primarily due to dynamic aspects of vLLM where LORA adapters can be loaded during runtime and where signature verification based on a vLLM-internal security policy provides more flexibility, including preventing the loading of any model that is not signed, or restricting vLLM to a set of models.
An example of a simple security policy is the following:

```
     {
        "policy": {
          "signatures": {
            "signers": {
              "signer1": {
                "verification_method": "sigstore",
                "identity": "Granite.Preview@ibm.com",
                "identity_provider": "https://sigstore.verify.ibm.com/oauth2"
              }
            },
            "models": {
              "/tmp/granite-4.0-tiny-preview/": {
                "signer": "signer1"
              }
            }
          }
        }
      }
```

This policy would allow only a single model in `/tmp/granite-4.0-tiny-preview` directory to be loaded and it must be signed with the 'sigstore' method by the given identity and identity_provider.

The matching of the vLLM command line model path against the policy is as follows:
- An exact match is attempted first by using the model path from the vLLM command line as key for the models dictionary in the policy (`/tmp/granite-4.0-tiny-preview`).
- If the exact match does not return a value, then the entries in the model paths in the models map are assumed to be regular expressions (must be prefixed with 'regex:') and are attempted to be matched against the model path from the vLLM command line. The first matching regular expression is used for selecting the signature verification parameters.

## Test Plan

The following new test cases have been added:
```
pytest tests/security/test_security_policy.py
```

The following test cases have been extended:
```
pytest tests/lora/test_lora_manager.py
```

## Test Result

TBD

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

stefanberger committed:
Add support for model signature verification on local models

Add support for model signature verification on local models. To do this,
pass a security policy to vLLM that describes the models by their local
(absolute) path and their expected signers. An example of such a policy
looks like this:

  {
    "policy": {
      "signatures": {
        "signers": {
          "signer1": {
            "verification_method": "sigstore",
            "identity": "Granite.Preview@ibm.com",
            "identity_provider": "https://sigstore.verify.ibm.com/oauth2"
          }
        },
        "models": {
          "/tmp/granite-4.0-tiny-preview": {
            "signer": "signer1"
          }
        }
      }
    }
  }

This policy would only allow the one shown model to be loaded and it must
pass signature verification using the sigstore method with the given
identity and identity_provider. The above policy assumes the default model
signature file model.sig.

Add test cases with valid and invalid policies.

With the above policy in /tmp/policy.json, it is now possible to verify the
signature on a signed model, such as an IBM Granite model:

cd /tmp
git clone https://huggingface.co/ibm-granite/granite-4.0-tiny-preview

vllm serve \
	--security-policy /tmp/policy.json
        /tmp/granite-4.0-tiny-preview/

Note that signature verification of models downloaded by vLLM from
Huggingface is not addressed in this patch.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
Improve reporting of error message if cause is known

Extend the reporting of an error over RPC by the cause of the error if it
is known. This then for example not only reports that the signature
verification failed but also the reason, such as when an unsigned file was
found.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
Return reason in error message as to why LoRA could not be loaded

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
logger: Add a logger for model_signing and adjust level when needed

Add a logger for the model_signing module and set its default log level
to WARNING. When the user wants to display the fingerprints of certificate
then temporarily set the logging level to INFO.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
temp: Have betterproto==2.0.0b6 installed by explicitly listing it

Since uv does not install the prerelease version 2.0.0b6 of betterproto,
explicitly list it among the common packages to install. Also list the
required version of sigstore so that this also gets installed.

The dependency tree is as follows:
 model-signing -> sigstore -> sigstore-protobuf-specs -> betterproto

The betterproto dependency is pulled in here:
https://github.com/sigstore/protobuf-specs/blob/main/gen/pb-python/pyproject.toml#L28

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

simon-mo reviewed (changes requested):
Thank you for submitting this PR. However, I don't this as in the scope of vLLM's responsibility. 

You are welcomed to leverage our plugin systems if this applies, or isolated as model loading process which can make plugin easier (we do intend to make model loader extensible).

stefanberger commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -75,6 +75,12 @@ def __init__(
         if self.log_stats:
             self.stat_logger = PrometheusStatLogger(vllm_config)
 
+        sc = self.model_config.security_config
+        if sc.model_signature_verification_needed(self.model_config.model):
```
Fixed.

stefanberger commented:
> Thank you for submitting this PR. However, I don't this as in the scope of vLLM's responsibility.

What are you referring to with 'this'. You mean verifying the signatures on AI models (or later on LORA adapters) or a security policy inside vLLM in general?

> 
> You are welcomed to leverage our plugin systems if this applies, or isolated as model loading process which can make plugin easier (we do intend to make model loader extensible).

If I was to rely on plugins I would probably have to set new plugin points with new interfaces where the security evaluation would have to happen. I am not sure how to go about the command line option for passing the security policy and how this should be handled.

stefanberger committed:
Enable signature verification on LoRA adapters

Enable signature verification on LoRA adapters loaded on startup as well
as those dynamically loaded during runtime. For this, extend the security
policy to allow for specifying LoRAs and their expected signers:

{
  "policy": {
    "signatures": {
      "signers": {
        "signer2": {
          "verification_method": "sigstore",
          "identity": "stefanb@us.ibm.com",
          "identity_provider": "https://sigstore.verify.ibm.com/oauth2"
        },
        "signer3": {
          "verification_method": "certificate",
          "certificate_chain": ["/home/stefanb/model-transparency/scripts/tests/keys/certificate/ca-cert.pem"]
        }
      },
      "models": {
        "/tmp/granite-3.3-8b-instruct/": {
          "signer": "signer2"
        }
      },
      "loras": {
         "/tmp/granite-3.3-8b-rag-agent-lib/query_rewrite_lora/" : {
             "signer": "signer3"
         },
         "regex:.*": {
             "signer": "signer2"
         }
      }
    }
  }
}

The above policy now enforces signature verification on the
query_rewrite_lora with the certificate method and all other LoRAs must
be signed with the sigstore method.

Make the SecurityPolicy class a 'dataclass' so it can be serialized. Add
a __post_init__ method it that will be called after de-serialization.

Extend the LoRARequest class to also carry the SecurityConfig that holds
the SecurityPolicy, which will be used for running signature verification
on the LoRA adapter.

Extend an existing test cases to test loading of unsigned and signed LoRA
adapters. For this, import a root and intermediate test CA and a signing
key for testing with the certificate/private key signing method.

With the above policy in /tmp/minimal-policy.json the LoRA adapter passed
on the command line as well as dynamically loaded LoRA adapters can have
their signatures enforced.

VLLM_ALLOW_RUNTIME_LORA_UPDATING=True vllm serve \
    --enforce-eager \
    --max-lora-rank 32 \
    /tmp/granite-3.3-8b-instruct/ \
    --enable-lora \
    --lora-modules \
       '{"name": "test", "path": "/tmp/granite-3.3-8b-rag-agent-lib/certainty_lora/"}' \
    --security-policy /tmp/minimal-policy.json

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
Instrument lora_filesystem_resolver for signature verification

When a LoRA adapter is loaded using the completions API, the
file_system_resolver plugin may be used to located the LoRA adapter.
Instrument the code path leading to the LoRA adapter to pass through the
SecurityConfiguation and attach it to the LoRARequest for later signature
verification of the LoRA adapter. Add the SecurityConfigration parameter
to the LoRAResolver interface and adjust all other classes implementing
this interface to accept the new parameter.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
docs: Add documentation about securty policy and signature verification

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

hmellor commented:
Closing as this feature has been deemed out of scope for vLLM.

stefanberger commented:
> Closing as this feature has been deemed out of scope for vLLM.

I just converted this PR to use the plugin system, as previously requested. However, it needs to add a new security plugin with different plugin points to the code base.

stefanberger commented:
Here's the branch with the security plugin: https://github.com/stefanberger/vllm/commits/signature-policy.4upstream/

stefanberger commented:
> Does enabling your plugin take 2k lines of code upstream?

I haven't looked, but is 2k the size of the patches right now? Some of it is documentation, other parts are test keys (in text format plus generating script), the rest is actual code and test cases with policies for testing etc. If needed I can cut out the text from the test keys and certs to make it smaller.

hmellor commented:
Plugins do not usually involve such large changes to upstream vLLM.

Documentation and testing of plugins should not really be upstreamed in this way, it kind of defeats the purpose of it being a plugin.

simon-mo commented:
I'm also interested in the business context, is IBM developing this as a product or research?

stefanberger commented:
> @stefanberger could the plugin be more like a generic "validate_model" hook? Then you can do whatever you want including sig verification, etc. All of the policy stuff could live in the plugin rather than the upstream code.

I can rename the 2 existing signature verification hooks to 'validate_model' and combine them into one. I will have it it have a parameter describing whether it is an AI model or a LoRA adapater so that a plugin can distinguish between the two, if necessary. There's currently another call to basically check whether validation of the model was asked for but was not done due to a code path that isn't instrumented, such as for example Huggingface downloads or code paths that I am not even aware of. Can we keep this call and maybe rename it to `model_needs_validation` or something like this?

Let me add the code for dealing with models downloaded from Huggingface to my repo. Because this code would have to go upstream as well so I can check those. I did not add this, yet, due to reasons related to the model signing library.

Would you want to have the SecurityPlugin also renamed, or are you fine with the name? Currently multiple plugins are supported and the first one that throws an Exception would get noticed by the user. I suppose we can leave it this way for now.

Further, I suppose you would not want the new plugin to be supported  with a command line option `--security-policy`. So I would have to resort to passing the filename of the policy using a plugin-specific environment variable?

> 
> The plugin itself could be tested in a self-contained way, and a unit test for the plugin mechanism itself should be trivial (just a dummy plugin which can either allow or disallow a particular model).

Yes, of course.
